audio/wav.go

```diff
@@ -0,0 +1,276 @@
+// If any of the parameters are 0, then it will be given the default 
+// values
+func NewWAVFromParams(params *WAVParams) *WAV {
+	// create a WAV from the given params
+	// use defaults from previous function if any value is 0
+	if params == nil {
+		return NewWAV()
+	}
+	if params.NumChannels == 0 {
+		params.NumChannels = DefaultNumChannels
+	}
+	if params.SampleRate == 0 {
+		params.SampleRate = DefaultSampleRate
+	}
+	if params.BitsPerSample == 0 {
+		params.BitsPerSample = DefaultBitsPerSample
+	}
+	if params.AudioData == nil || len(params.AudioData) == 0 {
+		params.AudioData = make([]byte, 0)
+	}
+	return &WAV{
+		NumChannels:   params.NumChannels,
+		SampleRate:    params.SampleRate,
+		AudioFormat:   DefaultAudioFormat,
+		BitsPerSample: params.BitsPerSample,
+		audioData:     params.AudioData,
+	}
+}
+
+// NewWAVFromData creates a WAV format struct from the given data buffer
+// The buffer is broken up into its respective information and that 
+// information is used to create the WAV format struct
+func NewWAVFromData(data []byte) (*WAV, error) {
+	// create a WAV from the given buffer.
+	// return error if len(data) < 44
+	// extract data from the data according to the spec: http://soundfile.sapp.org/doc/WaveFormat/
+
+	// find first data index
+	i := 4
+	for i < len(data) && data[i-4] != 'd' || data[i-3] != 'a' || data[i-2] != 't' || data[i-1] != 'a' {
+		i++
+	}
+
+	dataLen := len(data) - i
+	if dataLen <= 0 {
+		return nil, errors.Error{
+			Code:    "One",
+			Message: "Received WAV file with empty data",
+		}
+	}
+
+	// hOff is the header offset. It denotes the start of the header, which is
+	// 44 bytes before the data starts.
+	hOff := i - 44
+
+	numChannels := binary.LittleEndian.Uint16(data[hOff+22 : hOff+23])
+	sampleRate := binary.LittleEndian.Uint32(data[hOff+24 : hOff+27])
+	bitsPerSample := binary.LittleEndian.Uint16(data[hOff+34 : hOff+35])
+	audioData := data[i:]
+
+	return &WAV{
+		NumChannels:   numChannels,
+		SampleRate:    sampleRate,
+		AudioFormat:   DefaultAudioFormat,
+		BitsPerSample: bitsPerSample,
+		audioData:     audioData,
+	}, nil
+}
+
+// NewWAVFromReader takes in a reader and creates a new WAV format 
+// with the given information. 
+func NewWAVFromReader(reader io.Reader) (*WAV, error) {
+	b, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+	return NewWAVFromData(b)
+}
+
+// getRMS is a helper function used to calculate the RMS. This is called 
+// by TrimSilent which uses RMS to determine whether the sample of audio
+// is silent
+func getRMS(sampleSize uint16, audioData []byte) float64 {
+	sum := 0.0
+	for i := uint16(0); i < uint16(len(audioData)); i += sampleSize {
+		val := binary.LittleEndian.Uint64(audioData[i:(i + sampleSize)])
+		sum += math.Pow(2.0, float64(val))
+	}
+
+	return math.Sqrt(sum / (float64(len(audioData)/int(sampleSize))))
+}
+
+// TrimSilent is called on a WAV struct to trim the silent portions from the
```

Comment:
Mention that `padding` is specified in seconds.

Reply:
And that threshold is a decimal (between 0 and 1), relative to the maximum amplitude of the waveform.